contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -387,6 +450,126 @@ To deploy a containerized third-party CSI volume driver, it is recommended that
 
 Alternatively, deployment could be simplified by having all components (including external-provisioner and external-attacher) in the same pod (DaemonSet). Doing so, however, would consume more resources, and require a leader election protocol (likely https://git.k8s.io/contrib/election) in the `external-provisioner` and `external-attacher` components.
 
+#### Topology Representation in Node Objects
+Topology information will be represented as labels.
+
+Requirements:
+* Must adhere to the [label format](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set).
+* Must support different drivers on the same node.
+* The format of each key/value pair must match those in `PersistentVolume` and `StorageClass` objects, as described in the [Provisioning and Deleting](#provisioning-and-deleting) section.
+
+Proposal: `"com.example.topology/rack": "rack1"`
+The list of topology keys known to the driver is stored separately in the `CSINodeInfo` object.
+
+Justifications:
+* No strange separators needed, comparing to the alternative. Cleaner format.
+* The same topology key could be used across different sources (different storage plugin, network plugin, etc.)
+* Once NodeRestriction is moved to the newer model (see [here](https://github.com/kubernetes/community/pull/911) for context), for each new label prefix introduced in a new driver, the cluster admin has to configure NodeRestrictions to allow the driver to update labels with the prefix. Cluster installations could include certain prefixes for pre-installed drivers by default. This is less convenient compared to the alternative, which can allow editing of all CSI drivers by default using the “csi.kubernetes.io” prefix, but often times cluster admins have to whitelist those prefixes anyway (for example ‘cloud.google.com’)
+
+Considerations:
+* Upon driver deletion/upgrade/downgrade, stale labels will be left untouched. It’s difficult for the driver to decide whether other sources rely on this label.
+* During driver installation/upgrade/downgrade, controller deployment must be brought down before node deployment, and node deployment must be deployed before the controller deployment, because provisioning relies on up-to-date node information.
+* Topology keys inside `CSINodeInfo` must reflect the topology keys from drivers currently installed on the node. If no driver is installed, the collection must be empty.
```

Comment:
A similar race problem arises as well with node rolling upgrades - we could end up with a requisite topology that doesn't contain the selected node.

Reply:
Synced with @msau42 offline. Like she said, there is always going to be a chance the provisioner has out-of-date topology information on the node, but if the provisioner acts on that case, we are OK - it's not any worse than how CSI works today, since volumes could totally be provisioned in a volume where it's not accessible.

There's a few things we could do to minimize this possibility as much as possible. For example, if the driver info could be deleted from `CSINodeInfo` as soon as possible after a node driver deletion or crash, the provisioner could do the following:
* Never use that node as the source of truth for topology keys.
* Don't include that node in the aggregated requisite topology.
In addition, the scheduler could check `CSINodeInfo` and invalidate nodes that don't have the appropriate driver information as potential scheduling candidates.

Will update the doc